WIP: Support efficient SLA preservation in rolling updater

ironcladlou opened the pull request:
**WIP**

This is a WIP implementation of #3844 for discussion. It's a drastic refactor of the upstream rolling updater to support maintaining an SLA during updates. I'm keeping the PR open right now for design review and will likely close it and open another one later once all the design is hashed out.

ironcladlou commented:
cc @smarterclayton @kargakis 

This one's a doozy. It might be most productive to vet the test scaffolding and examine the test cases/add new ones to cover all the gaps. This is a massive refactor and the whole concept has many, many edge cases. If we agree the test scaffolding/DSL is solid then we can try and vet the algorithm/implementation in terms of test cases rather than debating the implementation at this point.

ironcladlou commented:
I also feel the need to defend the invasiveness of the refactor: I could not figure out how to express the test scenarios with the detail necessary using the fake clients/previous factoring. Hopefully the decomposed `Update` makes more sense structurally now even if the overall algo is more complex.

kargakis commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go:
```diff
@@ -368,253 +338,290 @@ func FindSourceController(r RollingUpdaterClient, namespace, name string) (*api.
+	// TODO: remove debugging output
+	fmt.Printf("Scaling %s down (current=%d, maxIncrement=%d, ready=%d, minReady=%d, oldReady=%d, newReady=%d, increment=%d)\n", oldRc.Name, oldRc.Spec.Replicas, maxIncrement, ready, minReady, oldReady, newReady, increment)
+	// We can't scale down without crossing the minimum. Try again later.
+	if increment == 0 {
```
Can't this be negative? What if there are less ready pods than what we want (minReady>ready) ?

ironcladlou replied:
Yeah, I think I need to make a test case for this.

kargakis replied:
I guess for now we can get away with not doing anything sophisticated about this but we should definitely know about it.

kargakis commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater_test.go:
```diff
@@ -328,199 +303,275 @@ Scaling foo-v2 up to 8
+			newRcExists:   false,
+			updatePercent: Percent(50),
+			expected: []interface{}{
+				up{2},
```
So to get this clear, the SLA is defined based on the difference set between the desired replicas of the first controller that will be scaled (newRc is we start up, oldRc if we start down) and (desired*update percentage), right?

ironcladlou replied:
More simply, the SLA is (old.replicas - (updatepercent*old.replicas)). With 100 old replicas, 10% update percent means keep 90 ready during the update. (If no percent is specified, max increment of 1 is used, so the SLA would be keep 99 running)

kargakis replied:
Got it, thanks

kargakis commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater_test.go:
```diff
@@ -328,199 +303,275 @@ Scaling foo-v2 up to 8
+			oldRc:         oldRc(1),
+			newRc:         newRc(0, 1),
+			newRcExists:   false,
+			updatePercent: Percent(-50),
```
Percent(-100) ?

kargakis commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go:
```diff
@@ -368,253 +338,290 @@ func FindSourceController(r RollingUpdaterClient, namespace, name string) (*api.
+		return false, nil
+	})
+	oldReady := ready[oldRc.Name]
+	newReady := ready[oldRc.Name]
```
you probably want to pass `ready[newRc.Name]` to `newReady` here

kargakis replied:
even better to `return ready[oldRc.Name], ready[newRc.Name], oldReady + newReady, err` but that's a nit

ironcladlou commented:
If you have a test case idea that can't be expressed with the existing test scaffolding, please pseudo-code it out here the way you wish you could express it.

ironcladlou commented:
Replaced the old continutations test with a new on in the scenarios table, added some more nuanced test cases, commented everywhere, moved some code around for clarity/flow, etc. The tests cases still need lots of intense scrutiny.

kargakis commented on Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go:
```diff
@@ -93,27 +78,400 @@ const (
+		increment = newIncrement
+
+		// Wait between down/up.
+		time.Sleep(config.UpdatePeriod)
```
Since we already wait in `scaleDown` for oldRc.Status.Replicas to sync with oldRc.Spec.Replicas, this feels redundant.

ironcladlou replied:
Good point. If we make efficient progress when possible and wait during scales, I'm not sure what these sleeps really accomplish anymore.

smarterclayton replied:
The original point of the sleeps was to give things a chance to stabilize
elsewhere in the system.  Such that a caller can add an arbitrary wait.
Also, you don't want to hotloop.  So you need some min wait period, and
then you also need the cooldown period.

On Tue, Jul 28, 2015 at 9:47 AM, Dan Mace <notifications@github.com> wrote:

> In
> Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/kubectl/rolling_updater.go
> <https://github.com/openshift/origin/pull/3908#discussion_r35648372>:
>
> > +			newRc = scaledRc
> > +			skipFirstUp = true
> > +		} // Otherwise, scale down/up.
> > +
> > +		// Scale down as much as possible while maintaining the minimum ready
> > +		// amount. The scale-down informs us as to how much we can safely scale up
> > +		// next interval.
> > +		scaledRc, newIncrement, err := r.scaleDown(newRc, oldRc, desired, maxIncrement, minReady, config)
> > +		if err != nil {
> > +			return err
> > +		}
> > +		oldRc = scaledRc
> > +		increment = newIncrement
> > +
> > +		// Wait between down/up.
> > +		time.Sleep(config.UpdatePeriod)
>
> Good point. If we make efficient progress when possible and wait during
> scales, I'm not sure what these sleeps really accomplish anymore.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3908/files#r35648372>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

kargakis commented:
We should move the discussion upstream so more people can weigh in on this.

ironcladlou commented:
>We should move the discussion upstream so more people can weigh in on this.

@smarterclayton what do you think? Since the previous patch was accepted, should I make a new PR upstream to continue this?

ironcladlou commented:
@smarterclayton would you mind making any new file comments on the main diff view rather than the commit iself? GitHub's UI makes commit level comments really difficult to work with, unfortunately.

smarterclayton commented:
Yes

On Tue, Jul 28, 2015 at 9:46 AM, Dan Mace <notifications@github.com> wrote:

> We should move the discussion upstream so more people can weigh in on this.
>
> @smarterclayton <https://github.com/smarterclayton> what do you think?
> Since the previous patch was accepted, should I make a new PR upstream to
> continue this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3908#issuecomment-125615691>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

ironcladlou commented:
@smarterclayton sigh... I'm pretty sure my commit amending/rebasing means all the comments we had going on a specific commit are now just gone forever. Sorry about that.

Let's use diff view comments henceforth to prevent further loss, since I'm continuously amending and rebasing during development.

ironcladlou commented:
I see now that the comments are still directly addressable in github if you have the link from emails, but they're now totally disassociated from this PR.

ironcladlou commented:
This was an important discussion:

>I think I see a fencepost problem. When we were talking before about whether we still needed the canary, it was in the context of a 10 or 20%. However, there is a reason for users to want to have a high % (like 50 or 100) but still do the check. If I'm a user, and I like the default behavior (checking for a ready pod), then setting 100% is worse. What I'm starting to think is that we should make canary a separate pre deploy action, called out separately. It's ok to continue down the path we have now, but we may have a "canary" path like a pre start hook or something.

If you used 100%, there would still need to be at least 1 new pod in a ready state before the old would be fully scaled down and the deployment completing successfully. Does that cover it?

ironcladlou commented:
What should I address before bringing this upstream?

smarterclayton commented:
On Jul 28, 2015, at 11:54 AM, Dan Mace <notifications@github.com> wrote:

This was an important discussion:

I think I see a fencepost problem. When we were talking before about
whether we still needed the canary, it was in the context of a 10 or 20%.
However, there is a reason for users to want to have a high % (like 50 or
100) but still do the check. If I'm a user, and I like the default behavior
(checking for a ready pod), then setting 100% is worse. What I'm starting
to think is that we should make canary a separate pre deploy action, called
out separately. It's ok to continue down the path we have now, but we may
have a "canary" path like a pre start hook or something.

If you used 100%, there would still need to be at least 1 new pod in a
ready state before the old would be fully scaled down and the deployment
completing successfully. Does that cover it?


It's more that you may not want to make N copies of something that will
fail.

We can come back to this later - it's not strictly required, we may just
want to have a "canary" hook.  The challenge is that you may not want the
canary in the rotation.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3908#issuecomment-125661898>.

ironcladlou commented:
Discussion moved upstream: https://github.com/GoogleCloudPlatform/kubernetes/pull/11942

ironcladlou commented:
Closing this in favor of the upstream PR. Will make a new origin PR when it's time to integrate.
